src/Couchbase.Lite.Shared/API/Database/Database.cs

```diff
@@ -871,14 +871,21 @@ internal void ResolveConflict([NotNull]string docID)
                         WriteLog.To.Database.I(Tag, "Resolving doc '{0}' (mine={1} and theirs={2})",
                             new SecureLogString(docID, LogMessageSensitivity.PotentiallyInsecure), localDoc.RevID,
                             remoteDoc.RevID);
-                        var resolvedDoc = ResolveConflict(localDoc, remoteDoc);
+                        conflictResolver = conflictResolver ?? new DefaultConflictResolver();
+                        var conflict = new Conflict(localDoc.IsDeleted ? null : localDoc, remoteDoc.IsDeleted ? null : remoteDoc);
+                        var resolvedDoc = conflictResolver.Resolve(conflict);
+                        if (resolvedDoc != null && resolvedDoc.Id != docID)
+                            throw new InvalidOperationException($"Resolved docID {resolvedDoc.Id} does not match with docID {docID}");
+                        if (resolvedDoc != null && resolvedDoc.Database != this)
```

Comment:
just to note, if the resolve function returns a newly created document, then the resolvedDoc.database might be null. which might cause this to throw.